### PR TITLE
correct plan for CTAS with volatile functions.

### PR DESCRIPTION
**Pull request by Sasasu**
follow up #10418.

CTAS has an additional code path that will add a replicated locus on the top of the current node.
but the code path lacks some volatile function check, which will produce an incorrect plan. The incorrect
plan will execute the volatile function more times than expected.

```
-- the old plan
> explain (costs off) create table rpt_ctas as select random() from generate_series(1, 10) distributed replicated;
+-------------------------------------+
| QUERY PLAN                          |
|-------------------------------------|
| Function Scan on generate_series    |
| Optimizer: Postgres query optimizer |
+-------------------------------------+

> explain (costs off) create table rpt_ctas as select a from generate_series(1, 10) a group by a having sum(a) > random() distributed replicated;
+---------------------------------------------------+
| QUERY PLAN                                        |
|---------------------------------------------------|
| HashAggregate                                     |
|   Group Key: a                                    |
|   Filter: ((sum(a))::double precision > random()) |
|   ->  Function Scan on generate_series a          |
| Optimizer: Postgres query optimizer               |
+---------------------------------------------------+

-- the new plan
> explain (costs off) create table rpt_ctas as select random() from generate_series(1, 10) distributed replicated;
+---------------------------------------------+
| QUERY PLAN                                  |
|---------------------------------------------|
| Broadcast Motion 1:3  (slice1; segments: 1) |
|   ->  Function Scan on generate_series      |
| Optimizer: Postgres query optimizer         |
+---------------------------------------------+

> explain (costs off) create table rpt_ctas as select a from generate_series(1, 10) a group by a having sum(a) > random() distributed replicated;
+---------------------------------------------------------+
| QUERY PLAN                                              |
|---------------------------------------------------------|
| Broadcast Motion 1:3  (slice1; segments: 1)             |
|   ->  HashAggregate                                     |
|         Group Key: a                                    |
|         Filter: ((sum(a))::double precision > random()) |
|         ->  Function Scan on generate_series a          |
| Optimizer: Postgres query optimizer                     |
+---------------------------------------------------------+
```

ORCA looks like not have this special optimization for replicated tables. in this case, ORCA's plan is correct.

related issues: #9737 #10419

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/test/regress/expected/rpt.out [hunk 1]
```diff
@@ -934,19 +934,39 @@ explain (costs off) select a from t_replicate_volatile union all select * from n
  Optimizer: Postgres query optimizer
 (6 rows)
 
+-- CTAS
+explain (costs off) create table rpt_ctas as select random() from generate_series(1, 10) distributed replicated;
```
Comment (kainwen) [inline src/test/regress/expected/rpt.out:938]: better mention github issue link here.

File: src/backend/cdb/cdbllize.c [hunk 2]
```diff
@@ -79,6 +79,7 @@
 #include "cdb/cdbmutate.h"
 #include "cdb/cdbutil.h"
 #include "optimizer/tlist.h"
+#include "optimizer/optimizer.h"
```
Comment (kainwen) [inline src/backend/cdb/cdbllize.c:82]: Why include this?

**Review by kainwen — APPROVED** [review]
Some comments. Others LGTM

**Comment by kainwen** [comment]
Another comment is better to add serial like test cases (from the related issue) into the case of the PR.

File: src/backend/cdb/cdbllize.c (hunk 2 above)
Comment (Sasasu) [reply src/backend/cdb/cdbllize.c:82]: add a comment and push -f

File: src/test/regress/expected/rpt.out (hunk 1 above)
Comment (Sasasu) [reply src/test/regress/expected/rpt.out:938]: #9737 is about INSERT or INSERT FROM and fixed by #10418
but this PR is only for CREATE AS, this issue is not reported in the community yet.

**Review by HaozhouWang — APPROVED** [review]

**Comment by JunfengYang** [comment]
More test cases are appreciated, please refer to https://github.com/greenplum-db/gpdb/pull/12226.

File: src/test/regress/expected/bfv_planner.out [hunk 3]
```diff
@@ -562,6 +562,107 @@ explain (costs off) select * from t_hashdist cross join (select * from generate_
...
+         Filter: (i < nextval('test_seq'::regclass))
+         ->  Function Scan on generate_series i
+ Optimizer: Postgres query optimizer
+(6 rows)
+
+create table t_rep2 as select i from generate_series(5, 15) as i group by i having i < nextval('test_seq') distributed replicated;
+select count(*) from gp_dist_random('t_rep2');
+ count 
+-------
+    33
```
Comment (Sasasu) [inline src/test/regress/expected/bfv_planner.out:607]: due to `HashAggregate` and `sequence cache` the result is different from 6X.

order of execution of the having clause is not guaranteed in `HashAggregate`

**Pull request merged**